source/validate_derivatives.cpp

```diff
@@ -0,0 +1,66 @@
+// Copyright (c) 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Validates correctness of derivative SPIR-V instructions.
+
+#include "validate.h"
+
+#include "diagnostic.h"
+#include "opcode.h"
+#include "val/instruction.h"
+#include "val/validation_state.h"
+
+namespace libspirv {
+
+// Validates correctness of derivative instructions.
+spv_result_t DerivativesPass(ValidationState_t& _,
+                          const spv_parsed_instruction_t* inst) {
+  const SpvOp opcode = static_cast<SpvOp>(inst->opcode);
+  const uint32_t result_type = inst->type_id;
+
+  switch (opcode) {
+    case SpvOpDPdx:
+    case SpvOpDPdy:
+    case SpvOpFwidth:
+    case SpvOpDPdxFine:
+    case SpvOpDPdyFine:
+    case SpvOpFwidthFine:
+    case SpvOpDPdxCoarse:
+    case SpvOpDPdyCoarse:
+    case SpvOpFwidthCoarse: {
+      if (!_.IsFloatScalarType(result_type) &&
+          !_.IsFloatVectorType(result_type)) {
+        return _.diag(SPV_ERROR_INVALID_DATA)
+               << "Expected Result Type to be float scalar or vector type: "
+               << spvOpcodeString(opcode);
+      }
+
+      const uint32_t p_type = _.GetOperandTypeId(inst, 2);
+      if (p_type != result_type) {
+        return _.diag(SPV_ERROR_INVALID_DATA)
+               << "Expected P type and Result Type to be the same: "
+               << spvOpcodeString(opcode);
+      }
+
```

Comment:
Deserves a comment here saying that the only other condition is that these instructions are only usable by the Fragment execution model. But that's must be checked elsewhere (after the entire module is read) since this instruction might be in a function that isn't itself an entry point.